src/Microsoft.Extensions.Configuration.AzureAppConfiguration/DnsClient/DnsProcessor.cs

```diff
@@ -0,0 +1,390 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration.DnsClient
+{
+    internal class DnsProcessor
+    {
+        private const ushort SrvType = 33;  // SRV DNS type code
+        private const ushort InClass = 1;  // IN DNS class code
+        private const ushort OptRrType = 41;  // OPT DNS type code
+        private const int RetryAttempt = 3;
+
+        private const string OriginSrvPrefix = "_origin._tcp";
+        private string AlternativeSrvPrefix(ushort index) => $"_alt{index}._tcp";
+
+        public async Task<IReadOnlyCollection<SrvRecord>> QueryAsync(IPEndPoint endpoint, string query, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var udpClient = new UdpClient(endpoint.AddressFamily);
+
+            var resultRecords = new List<SrvRecord>();
+
+            try
+            {
+                using var callback = cancellationToken.Register(() =>
+                {
+#if !NET45
+                    udpClient.Dispose();
+#else
+                    udpClient.Close();
+#endif
+                });
+
+                var originRecords = await QueryAsyncInternal(endpoint, $"{OriginSrvPrefix}.{query}", udpClient, cancellationToken).ConfigureAwait(false);
+                string originHost = query;
+                if (originRecords != null && originRecords.Count > 0)
+                {
+                    originHost = originRecords.First().Target;
+                }
+                else
+                {
+                    // If can't get any records from _origin query, we should return;
+                    return resultRecords;
+                }
+
+                IReadOnlyCollection<SrvRecord> altRecords = null;
+                ushort index = 0;
+
+                while (true)
+                {
+                    altRecords = await QueryAsyncInternal(endpoint, $"{AlternativeSrvPrefix(index)}.{originHost}", udpClient, cancellationToken).ConfigureAwait(false);
+
+                    if (altRecords == null || altRecords.Count == 0)
+                    {
+                        break;
+                    }
+
+                    resultRecords.AddRange(altRecords);
+
+                    index++;
+                }
+
+                return resultRecords;
+            }
+            catch (SocketException se) when (se.SocketErrorCode == SocketError.OperationAborted)
+            {
+                throw new TimeoutException();

```

Comment:
Seems like we should be catching these thrown exceptions like `TimeoutException` and `DnsResponseException` in `TryRefreshAsync` and wherever else we might be calling this

Reply:
`TryRefreshAsync` would not call this method directly, it calls the `SrvLookupClient.QueryAsync()` method, it has catch all exceptions and logging warning, does it make sense to you?

Reply:
My fault, I didn't see that there were 2 `QueryAsync` methods. Makes sense!